samcli/commands/local/cli_common/invoke_context.py

```diff
@@ -138,17 +186,54 @@ def __enter__(self):
                 "Running AWS SAM projects locally requires Docker. Have you got it installed and running?"
             )
 
+        # initialize all lambda function containers upfront
+        if self._warm_containers and not self._lazy_containers_initializing:
+            self._initialize_all_functions_containers()
+
         return self
 
     def __exit__(self, *args):
         """
-        Cleanup any necessary opened files
+        Cleanup any necessary opened resources
         """
 
         if self._log_file_handle:
             self._log_file_handle.close()
             self._log_file_handle = None
 
+        if self._warm_containers:
+            self._clean_running_containers_and_related_resources()
+
+    def _initialize_all_functions_containers(self):
+        """
+        Create and run a container for each available lambda function
+        """
+        LOG.info("Initializing the lambda functions containers.")
+
+        async def initialize_function_container(function):
+            function_config = self.local_lambda_runner.get_invoke_config(function)
+            self.lambda_runtime.run(None, function_config, self._debug_context, None)
+
+        try:
+            loop = asyncio.get_event_loop()
+            loop.run_until_complete(
+                asyncio.gather(
+                    *[initialize_function_container(function) for function in self._function_provider.get_all()]
+                )
+            )
```

Comment:
You can use previously created `AsyncContext` in following place. https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/utils/async_utils.py#L87

This will help and isolate usage of asyncio, and we can upgrade or change it in the future if we needed to.

Reply:
Done